cloud/cloud_scheduler.cc

```diff
@@ -67,18 +70,33 @@ class CloudSchedulerImpl : public CloudScheduler {
 
 // Implementation of a CloudScheduler that keeps track of the jobs
 // it scheduled.  Only cleans up those jobs on exit or cancel.
-class LocalCloudScheduler : public CloudScheduler {
+class LocalCloudScheduler : public CloudScheduler,
+                            std::enable_shared_from_this<LocalCloudScheduler> {
+  struct PrivateTag {
+    PrivateTag() = default;
+  };
  public:
-  LocalCloudScheduler(const std::shared_ptr<CloudScheduler>& scheduler,
+  static std::shared_ptr<CloudScheduler> Create(
+      const std::shared_ptr<CloudScheduler>& scheduler, long local_id) {
+    return std::make_shared<LocalCloudScheduler>(
+      PrivateTag(),
+      scheduler,
+      local_id);
+  }
+
+  LocalCloudScheduler(PrivateTag,
+                      const std::shared_ptr<CloudScheduler>& scheduler,
                       long local_id)
       : scheduler_(scheduler),
         next_local_id_(local_id),
         shutting_down_(false) {}
   ~LocalCloudScheduler() override {
-    {
-      std::lock_guard<std::mutex> lk(job_mutex_);
-      shutting_down_ = true;
-    }
+    TEST_SYNC_POINT(
+        "LocalCloudScheduler::~LocalCloudScheduler:BeforeCancelJobs1");
+    TEST_SYNC_POINT(
+        "LocalCloudScheduler::~LocalCloudScheduler:BeforeCancelJobs2");
+    std::lock_guard<std::mutex> lk(job_mutex_);
```

Comment:
Acquiring a mutex in the destructor is generally an antipattern - it implies that other calls can race with the destructor, and if that is the case, then that means those calls could also run after the destructor, resulting in use-after-free. Are you sure this code is correct?

Reply:
Let me rephrase - if we acquire this lock, then another thread blocks on it, we release the lock at the end of this destructor and promptly destroy it. The other thread will then wake up and start reading deallocated memory.

Reply:
What function do you think can race with the destructor? If you are worried about DoErase, remember we have a locked shared pointer there , so the destructor will not run.

Reply:
ahh good point!  yes, I'm worried about DoErase.

So there is even no need to lock the mutex here. Thanks for pointing it out!